### ISSUE TITLE
Simultaneous async git operations - best practice?

### ISSUE THREAD

**Issue by Nils-Kolvenbach** [desktop/dugite#474]
This might be too specific to my project so feel free to close this but maybe it's more common and someone has a best practice or some ideas for me?

I ran into multiple `GitError` messages in my logs like `Git (2.32.0) command [COMMAND] failed with code 128 and message: fatal: Unable to create '[PATH].git/index.lock': File exists. Another git process seems to be running in this repository, e.g. an editor opened by 'git commit'. Please make sure all processes are terminated then try again. If it still fails, a git process may have crashed in this repository earlier: remove the file manually to continue.`

From my understanding what the '[PATH].git/index.lock' file does is to ensure that simultaneous changes to your local git repo do not occur, as this would cause your git repo to be in an indeterminate state. The index.lock file prevents changes to your local repository from happening from outside of the currently running git process so as to ensure multiple git processes are not altering or changing the same repository internals at the same time.

So if I have multiple async functions executing simultaneously (which is very common in my code) this would result in the aforementioned error. Some ideas coming to my mind how to solve this:
1. Every git command is queued so that they execute in order and not simultaneously. Unfortunately with this approach I cannot return the `Promise<IGitResult>` to the correct method that executed that command / the method is not waiting for the command to execute. Or is there a queue that could do that for me?
2. Before the execution of a git command I check if the '[PATH].git/index.lock' file exists and if so, delay the execution / or rerun the command on error code 128 until it resolves. But with this the order of my git commands is not respected, e.g. resulting in a `commit` being executed before the `add` command.
3. Before the execution of a git command I remove the '[PATH].git/index.lock' file if it exists. I do not know what could really happen if I do so but it's probably a bad idea.

I guess my question is simply if anybody has run into a similar issue or knows how I could handle this better than the ideas I came up with. Any help is much appreciated!

**Comment by niik** [comment]
Hey @Nils-Kolvenbach, thanks for reaching out. We (the GitHub Desktop team) has tried various solutions to work around this over the years see for example https://github.blog/2015-10-20-git-concurrency-in-github-desktop/ (this was prior to our Electron rewrite).

That approach ended up causing almost as much problems as not having it due to subtle multi-threading race conditions and deadlocks so when we rewrote GitHub Desktop we opted against any holistic solution and chose instead to be mindful of which Git commands we run in parallel and which we don't. Unfortunately there's no straightforward answer to what can safely be run alongside other commands. Listing branches can appear to be safe but isn't necessary if something like `pull` or `fetch` is happening concurrently for example. And far from all operation are guarded by lock files like in your example above. Trial and error as well as reading Git source code to understand what a command is going to do under the hood has proven to be the best approach for us.

For the most part running Git commands sequentially and awaiting each result before moving forward as a default mode and being explicit about when we run multiple in parallel solves the overwhelming majority of these issues for us but the ability to do so is obviously dependent on the application model. We have our own locking model for some long-running operations (like pull, fetch, push) that ensure we only run one of them per repo at a time (see https://github.com/desktop/desktop/blob/50c7fdf54ff8735ca4fe0f24b6d370e1c64dcf16/app/src/lib/stores/app-store.ts#L4079-L4127).

> From my understanding what the '[PATH].git/index.lock' file does is to ensure that simultaneous changes to your local git repo do not occur

It's a bit more specific than that. It's not a general purpose lock for the entire repository rather it ensures that only one Git process writes to the index (also known as the staging area, or cache) at one time.

> * Every git command is queued so that they execute in order and not simultaneously. Unfortunately with this approach I cannot return the `Promise<IGitResult>` to the correct method that executed that command / the method is not waiting for the command to execute. Or is there a queue that could do that for me?

You could certainly implement a synchronization mechanism with something like [p-queue](https://www.npmjs.com/package/p-queue) that would run only one command at a time while still returning promises to calling functions just like you do today. It'd look something like this

```ts
const queue = new PQueue({ concurrency: 1 })

const queuedExec = (
  args: string[],
  path: string,
  options?: IGitExecutionOptions
) => queue.add(() => GitProcess.exec(args, path, options))
```

This is the safest approach but also the least efficient since you're never running any commands in parallel. Note that if you're working with multiple repositories you'd run on queue per repo.

> * Before the execution of a git command I check if the '[PATH].git/index.lock' file exists and if so, delay the execution / or rerun the command on error code 128 until it resolves. But with this the order of my git commands is not respected, e.g. resulting in a `commit` being executed before the `add` command.

You can't trust the result of checking for the `index.lock` as it may be created in between you checking for it and executing the next command. Also note that exit code 128 is a generic exit code that can signal a number of failures distinct from lock file issues.

> * Before the execution of a git command I remove the '[PATH].git/index.lock' file if it exists. I do not know what could really happen if I do so but it's probably a bad idea.

Yeah, you'll be in a world of pain if you go down this route 😛

There's also options like [--no-optional-locks](https://git-scm.com/docs/git#Documentation/git.txt---no-optional-locks) which can be used in very specific scenarios to reduce contention on locks (git status being the primary one).

---

I hope that helps you onwards a bit but as you've already noted this is out of scope for this project so I'm going to close the issue. Concurrency management can only be applied correctly when taking the application model into account and a library can't do that. We could provide a queue solution like the one above as a last-resort option but even that won't save you if there's multiple processes operating in the repository (like GitHub Desktop and VS Code) and it's fairly straightforward to implement in the application itself should you need to.
